### ISSUE TITLE
Enable Mutually Aware NSEs to overlap in IPAM/Routes

### ISSUE THREAD

**Issue by edwarnicke** [networkservicemesh/api#121]
Network Service Mesh's exclude prefix mechanism is designed to protect the IPAM/Route Assignments an NSE makes from colliding with the prefixes already in use by the workloads environment (like the K8s cluster prefixes) *and* with the prefixes used by other NSEs to which the workload is connected.

Normally this means that if a workload connects to NSE-1 which assigns the workload an IP address and some routes, those should be added to the exclude prefix list for Requests to future NSEs.

One side effect of this is that it preclude the same IP address from being assigned to the interfaces from two different NSEs.

This is done because NSEs are presumed to be mutually ignorant of one another.  There is no safe way to allow IP space overlap between two mutually ignorant NSEs.

In https://github.com/networkservicemesh/api/issues/119 a really good use case was raised for mutually *aware* NSEs in the form of multihoming a loadbalancer that assigns the same VIP address coming from multiple NSEs.  It's important to support this use case.

In order for NSM to safely allow mutually aware NSEs to have overlapping IP space, it is necessary for NSEs to have a way to credibly signal that they are part of a group of mutually aware NSEs.

My initial idea for accomplishing this would be to add a field to ConnectionContext to hold a key unique to each 'awareness group' to indicate membership in that group.

There is some additional thought necessary around how those 'awareness group' ids associate to exclude prefixes so that the correct entities in the chain (like forwarders) can know its OK to allow overlapping prefixes, even if different  connections are handled by different forwarwders.

@szvincze Thoughts?

**Comment by denis-tingaikin** [comment]
@edwarnicke , @szvincze Is this ready to go?

**Comment by edwarnicke** [comment]
@denis-tingaikin I think so yes.

**Comment by Mixaster995** [comment]
Prepared decomposition for this issue:
- [ ] https://github.com/networkservicemesh/api/issues/123
- [ ] https://github.com/networkservicemesh/sdk/issues/1203
- [ ] https://github.com/networkservicemesh/cmd-nse-icmp-responder/issues/358

**Comment by denis-tingaikin** [comment]
@edwarnicke Could you have a look?

**Comment by denis-tingaikin** [comment]
@fkautz cc

**Comment by edwarnicke** [comment]
> Main question is: is it correct, that grouping will be NSE’s property, not NS’s property? We have grouped endpoints, unrelated to which NS they providing?

Yes, grouping will be NSE's property, not NS's.

**Comment by edwarnicke** [comment]
> ## Problem
>`client` a priory doesn’t know to what exact endpoint it will go and can’t correctly manage which exlcluded prefixes it should add to request. It knows NS, but in the same NS could be different groups(or no groups) of NSEs.

Let me rephrase the problem slightly (and let me know if I've missed an important detail in doing so):

# Problem

How do NSE's know which of the excluded_prefixes to respect and which to ignore?

**Comment by edwarnicke** [comment]
@denis-tingaikin Does this get to the root of the issue?

**Comment by denis-tingaikin** [comment]
@edwarnicke 

Hmm, In general, yes, but let me clarify a few points.

1. We're adding exclude prefixes from the NSE response on the NSC side see at https://github.com/networkservicemesh/sdk/blob/main/pkg/networkservice/common/excludedprefixes/client.go#L35
2. Since we do point 1, we can not know which of the excluded_prefixes to respect and which to ignore before Request to NSE.

**Comment by denis-tingaikin** [comment]
We also can add a configurable env `GroupBy` and have it by default `groupID` to let NSC know that it can have the same IPs for the connection.

**Comment by edwarnicke** [comment]
OK... having talked about this a bit in this weeks community meeting, it seems we need something fairly generic.  In particular, we need something that will allow mutual awareness between different Network Services.

Rather than getting tied up in what to name things up front... lets thing about what we logically have.

We need a way to 'bin' various requested network service connections into groups that are considered 'mutually aware'.

Would it make sense to do something like NSM_AWARENESS_GROUPS="[${nsurl},${nsurl}],[${nsurl},${nsurl}]" where each 'list' is basically a list of nsurls that don't need to include each others prefixes in exclude prefixes?

**Comment by denis-tingaikin** [comment]
> Would it make sense to do something like NSM_AWARENESS_GROUPS="[${nsurl},${nsurl}],[${nsurl},${nsurl}]" where each 'list' is basically a list of nsurls that don't need to include each others prefixes in exclude prefixes?

Yes, it definitely solves the problem. But we should pay attention that we should use for grouping some labels on the NSC and NSE sides. As said on the community call not all NSEs from the network services can be joined to the group. 


So, I would suggest doing next things:

1. Add NSM_AWARENESS_GROUPS for clients
2. Pass to nsurl some label for example: "?color=red" (or any other for grouping)
3. Use the same label as at step2 on NSE registration.

**Comment by edwarnicke** [comment]
@denis-tingaikin If I understand correctly you are suggesting an optional addition practice of using labels to make sure of mutual awareness by only a subset of NSEs?  What I think you are suggesting would just work, but would not be required.

It also looks like it doesn't require additional implementation in code beyond  NSM_AWARENESS_GROUPS correct?

**Comment by denis-tingaikin** [comment]
@edwarnicke yes, we don't need to do any modifications to the code. I think we need to use the labels in the example for this feature to show how to get a subset of nses

**Comment by edwarnicke** [comment]
@denis-tingaikin As long as we are clear in the text for the example that use of labels is optional, its good to show folks how to do it :)

**Comment by denis-tingaikin** [comment]
Finally, I very like how is it going. NSM_AWARENESS_GROUPS approach looks super flexible.

**Comment by edwarnicke** [comment]
And super simple

**Comment by edwarnicke** [comment]
Is this finished?  If so can we close it?

**Comment by denis-tingaikin** [comment]
Seems like it's done.
